# Deep-TEIRI 

A developed pipeline to identify TE-initiated transcripts by integrating both DNA sequence and RNA-seq coverage information using a deep-learning model.

This repository contains the pipeline (script/) and deep-learning model for TSS prediction (Model/).

## Software requirements

Python 3.12.3, BEDTools, gffread, numpy, pandas, torch, selene_sdk, pyBigWig, scipy, matplotlib, tabix

## Usage

### Input files 

```
GTF files: 
    GTF files which could be generated by stringtie from short-read RNA-seq.
BW files:
    Two compressed BW format files, containing coverage information (CPM normalized) for uniquely mapped and multi-mapped RNA-seq reads, which could be generated by deepTools.
TE reference bed: 
    TE annotation in BED format (at least 6 fields: chrom, chromStart, chromEnd, ID, name, and strand).
Reference GTF: 
    reference GTF file for gene annotations.
Reference fasta file:
    A FASTA file containing the nucleotide sequences of chromosomes or scaffolds.
Chromosome length file:
    A text file containing the lengths of chromosomes.
```

### TEIRI_correct.py

TEIRI_correct.py predicted the TSSs for each assembled transcript using the deep-learning model.

```
usage: TEIRI_correct.py [-h] [--input INPUT] [--short_path SHORT_PATH] [--chr_len CHR_LEN] [--window_length WINDOW_LENGTH] [-g REF_FA] [-m MODEL_PATH] [-p PREFIX]

TEIRI_correct

options:
  -h, --help            show this help message and exit
  --input INPUT         Prefix of input files (required)
  --short_path SHORT_PATH
                        Path to input files (default: '.')
  --chr_len CHR_LEN     Chromosome length file (required)
  --window_length WINDOW_LENGTH
                        Window size for TSS prediction (default:10000)
  -g REF_FA, --ref_fa REF_FA
                        Reference fasta file (required)
  -m MODEL_PATH, --model_path MODEL_PATH
                        Path to the deep-learning model (required)
  -p PREFIX, --prefix PREFIX
                        Prefix for ouput files (default: TEIRI)
```

### TEIRI_merge.py

TEIRI_merge.py is designed for the identification of TE-initiated transcripts and transcript merge using a reference-guided approach across various samples.

```
usage: TEIRI_merge.py [-h] [-i GTF_LIST] [-r REFERENCE_GTF] [--TE_anno TE_ANNO] [--tss_window TSS_WINDOW] [--max_tss MAX_TSS]
                      [--min_NGS_samples MIN_NGS_SAMPLES] [--max_transcripts MAX_TRANSCRIPTS]
                      [--trunctated_exclude TRUNCTATED_EXCLUDE] [--min_transcript_length MIN_TRANSCRIPT_LENGTH]
                      [--threshold THRESHOLD] [-p PREFIX]

TEIRI_merge

options:
  -h, --help            show this help message and exit
  -i GTF_LIST, --gtf_list GTF_LIST
                        A text file with a list of GTF files (required)
  -r REFERENCE_GTF, --reference_gtf REFERENCE_GTF
                        Reference genome annotation in GTF format (required)
  --TE_anno TE_ANNO     TE annotation in BED format (required)
  --tss_window TSS_WINDOW
                        The window size used to calculate the weight score (default: 50)
  --max_tss MAX_TSS     Maximum number of TSS picked per first exon (default: 1)
  --min_NGS_samples MIN_NGS_SAMPLES
                        The min NGS ratio supporting the first exon (default: 10)
  --max_transcripts MAX_TRANSCRIPTS
                        The max counts of transcripts for a TE-initiated RNA (default: 10)
  --trunctated_exclude TRUNCTATED_EXCLUDE
                        Truncated transcripts were excluded, as they may represent fragments of the full-length transcript
                        (default: True)
  --min_transcript_length MIN_TRANSCRIPT_LENGTH
                        The min transcript length (default: 200)
  --threshold THRESHOLD
                        The threshold of weight score for TSS (default: 1)
  -p PREFIX, --prefix PREFIX
                        Prefix for output file (default: TEIRI)
```

### TEIRI_consolidate.py

TEIRI_consolidate.py is designed for the consolidation and annotation of TE-initiated transcripts.

```
usage: TEIRI_consolidate.py [-h] [-i GTF_LIST] [-r REFERENCE_GTF] [--TE_anno TE_ANNO]
                            [--tss_merge_distance TSS_MERGE_DISTANCE] [--min_exon_length MIN_EXON_LENGTH] [-p PREFIX]

TEIRI_consolidate

options:
  -h, --help            show this help message and exit
  -i GTF_LIST, --gtf_list GTF_LIST
                        A text file with a list of GTF files (required)
  -r REFERENCE_GTF, --reference_gtf REFERENCE_GTF
                        Reference genome annotation in GTF format (required)
  --TE_anno TE_ANNO     TE annotation in BED format (required)
  --tss_merge_distance TSS_MERGE_DISTANCE
                        The min distance to merge TE-derived TSSs of different GTF files (default: 50)
  --min_exon_length MIN_EXON_LENGTH
                        The min length of first exon (default: 20)
  -p PREFIX, --prefix PREFIX
                        Prefix for output file (default: TEIRI)
```
